Standing up the YES transportation tool

el-mapache opened the pull request:
This PR adds the basic building blocks of the new YES transportation tool. There are a lot of upfront changes and I'd like to get eyes / approval on it before I add much more!

## Additions

- Entry point for app in `base.py` settings
- Folders in `unprocessed/apps` for static content and in `cfgov` for templates and python files
- Minimal python files, including a migration, for a new page type. This code could probably be removed in a new PR. I'm not sure if we want to introduce a migration before we know all the content the new page type might be storing.
- Templates for the `budget` and `transportation tool` sections. The templates collectively represent a single page.
- some CSS extensions
- JS files to cover a few views, basic state management, and tests


## Removals

-

## Changes

-

## Testing

Apologies if any of these steps are off base, I'm not super familiar with the exact order of things in wagtail!

1. Run migration included in this PR
2. Add a new child page to  `/practitioner-resources/resources-youth-employment-programs` called 'transportation-tool', with a page type of `TransportationTool`
3. Build assets and start the app
4. You _should_ be able to see the current page and interact with the form fields (most only store internal state at this point)

## Screenshots


## Notes

- We don't want the page to be live at this point. This PR is for adding visibility to the code, and to get something in the repo before too much more content is added.

- We don't have 'final' designs so usability without JS or CSS isn't possible yet; this is more about standing something up

## Todos

- Modify`BudgetFormView` to use `InputView` component in a future PR

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

## Testing checklist

### Browsers

- [ ] Chrome on desktop
- [ ] Firefox
- [ ] Safari on macOS
- [ ] Edge
- [ ] Internet Explorer 9, 10, and 11
- [ ] Safari on iOS
- [ ] Chrome on Android

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

el-mapache commented on test/unit_tests/apps/youth-employment-success/js/budget-form-view-spec.js:
```diff
@@ -0,0 +1,102 @@
+      
+      <div class="content-l_col content-l_col-2-3">
+        <span><b>(-)</b></span>
+        <input type="text" name="money-spent" class="a-text-input o-yes-budget-spent">
```
I don't think we need labels in specs?

el-mapache commented on cfgov/unprocessed/apps/youth-employment-success/js/route-option-view.js:
```diff
@@ -0,0 +1,106 @@
+  );
+
+  /**
+   * Updates form state from child <input text> nodes
```
Looks like a bot regex error

el-mapache commented on cfgov/unprocessed/apps/youth-employment-success/js/input-view.js:
```diff
@@ -0,0 +1,78 @@
+
+/**
+ * InputView
+ * Generic view for <input> elements
```
Looks like a bot regex error

el-mapache commented on cfgov/unprocessed/apps/youth-employment-success/js/input-view.js:
```diff
@@ -0,0 +1,78 @@
+const defaultProps = {
+  type: 'text'
+};
+const NODE_MISSING_ERROR = 'InputView expects to be initialized with an <input> node matching the supplied `type` prop';
```
Looks like a bot regex error

el-mapache commented on test/unit_tests/apps/youth-employment-success/js/input-view-spec.js:
```diff
@@ -0,0 +1,74 @@
+
+const HTML = `
+  <div>
+    <input type="text" name="my-input">
```
Don't believe specs need VO support

el-mapache commented on cfgov/youth_employment_success/jinja2/youth-employment-success/budget-form.html:
```diff
@@ -0,0 +1,68 @@
+        </div>
+      </div>
+      <div class="content-l_col content-l_col-2-3">
+        <input type="text" name="money-earned" class="a-text-input o-yes-budget-earned">
```
This input's label is correctly spoken using VoiceOver on my mac.

el-mapache commented on cfgov/youth_employment_success/jinja2/youth-employment-success/budget-form.html:
```diff
@@ -0,0 +1,68 @@
+      
+      <div class="content-l_col content-l_col-2-3">
+        <span><b>(-)</b></span>
+        <input type="text" name="money-spent" class="a-text-input o-yes-budget-spent">
```
This input's label is correctly spoken using VoiceOver on my mac.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/input-view.js:
```diff
@@ -0,0 +1,78 @@
+ */
+function InputView( element, props = {} ) {
+  const finalProps = {
+    ...defaultProps,
```
What's your browser support requirements for this app? Spread operator doesn't have IE11 support I believe, which we do deliver JS to. This likely will not be handled by webpack/babel. See the comment in [webpack.config.js ](https://github.com/cfpb/cfgov-refresh/pull/5176/files#diff-33ad45ad5b983390a0ca720f55f3d49eR45) `Use useBuiltIns: 'usage' and set `debug: true` to see what scripts require polyfilling.`. You can change the settings that way and when you build the script you'll see output in the Terminal of what should be polyfilled. We don't automatically include babel-polyfill because it adds polyfills for parts of third party dependencies that we never utilize, so adds unnecessary bloat. Sometimes it's easier to just use older style code structures for compatibility.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/money.js:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * Converts an input string into a scaled dollar value, or zero.
+ * @param {String} dollars Amount as string
```
Trivial, should be `{string}` and `{number}` in the jsdoc types since you're taking and returning a string and number primitive, respectively, not an instance of String and Number.

el-mapache replied:
I actually did not realize that, thanks!

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/reducers/budget-reducer.js:
```diff
@@ -0,0 +1,42 @@
+import keymirror from 'keymirror';
+
+const initialState = {
```
Will this be the same state as in `BudgetFormView`? If so, does this perhaps indicate state should be spun out in its own module that can be shared? It may be too small to warrant spinning out.

el-mapache replied:
This state is going to replace the local state in `BudgetFormView`, since the tool will need that data is a few different places. I haven't actually gone through and wired up the reducer logic yet :/

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/reducers/budget-reducer.js:
```diff
@@ -0,0 +1,42 @@
+  spent: ''
+};
+
+const actionTypes = keymirror( {
```
Worth it to create a dependency here on `keymirror`? Seems like you could have:

```
const actionTypes = {
  UPDATE_EARNED: 0,
  UPDATE_SPENT: 1
};
```

Or whatever the value you need it to be. E.g. keymirror would be creating this, right?

```
const actionTypes = {
  UPDATE_EARNED: 'UPDATE_EARNED',
  UPDATE_SPENT: 'UPDATE_SPENT'
};
```

Seems like it may be clearer to set the value directly.

el-mapache replied:
Yes, I just like not having to manually type out the string, in case of typos. I thought projects had their own `node_modules`? Or do dependencies still bloat the main js?

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/reducers/budget-reducer.js:
```diff
@@ -0,0 +1,42 @@
+export const updateEarnedAction = updateAction( actionTypes.UPDATE_EARNED );
+export const updateSpentAction = updateAction( actionTypes.UPDATE_SPENT );
+
+function budgetReducer( state = initialState, action ) {
```
Function needs jsdoc comment.

anselmbradford replied:
Technically type is `Object` (not a primitive)

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/route-option-view.js:
```diff
@@ -0,0 +1,106 @@
+  );
+
+  /**
+   * Updates form state from child <input text> nodes
```
Should this say `<input> text nodes`?

el-mapache replied:
Maybe? I was just short-handing `<input type="text">` and am willing to change it to whatever you think is more clear 😄

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/route-option-view.js:
```diff
@@ -0,0 +1,106 @@
+   * @param {Object} event DOM event object
+   */
+  function _setQuestionResponse( name, event ) {
+    state = {
```
Could this be `state[name] = event.target.value;`?

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/route-option-view.js:
```diff
@@ -0,0 +1,106 @@
+
+  /**
+   * Updates state from child <input checkbox> nodes
+   * @param {*} _ unused
```
Why's this included if it's unused?

el-mapache replied:
The linter was complaining about this param not being noted in the jsdoc

anselmbradford replied:
I mean why's the param there at all?

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/route-option-view.js:
```diff
@@ -0,0 +1,106 @@
+      safeSelected.push( value );
+    }
+
+    state = {
```
Could this be `state.selectedTransportation = safeSelected;`? Is the spread doing something special here that I'm misssing?

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/util.js:
```diff
@@ -0,0 +1,56 @@
+  const finalState = {};
+  let hasChanged = false;
+
+  for ( let i = 0; i < reducersKeys.length; i++ ) {
```
Could be `for ( let i = 0, len = reducersKeys.length; i < len; i++ ) {`, though I'm not totally certain that's more efficient.

el-mapache replied:
The v8 optimizing compiler, at least, will automatically cache the loop length as a variable when the code. I'm not sure how other compilers treat it, but I think the `.length` property access is probably safe.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/util.js:
```diff
@@ -0,0 +1,56 @@
+    const reducer = reducers[key];
+
+    /**
+     * if we didnt already get a composite state object, e.g. if this is
```
TRIVIAL: Follow proper grammar in code comments (capitalization, ending period, etc.), so it's easier at a glance to see if a comment has been inadvertently truncated, for instance.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/util.js:
```diff
@@ -0,0 +1,56 @@
+     */
+    const previousState = state ? state[key] : state;
+    let argumentsToApply = [ action ];
+    argumentsToApply = [ previousState, ...argumentsToApply ];
```
Could this line be combined into:
```
let argumentsToApply = [ previousState, action ];
```
Or am I missing something on what the spread is doing here?

el-mapache replied:
Nope! I wrote this pretty quickly, that can be changed for sure.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/util.js:
```diff
@@ -0,0 +1,56 @@
+  return hasChanged ? finalState : state;
+};
+
+export const combineReducers = ( reducers = {} ) => {
```
Lot of nested logic here. Okay, but would be clearer if the complexity was reduced. Food for thought.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/yes-store.js:
```diff
@@ -0,0 +1,17 @@
+import Store from '../../../js/organisms/MortgagePerformanceTrends/stores/store';
```
Hmm okay for now, but maybe we should move this store to a `utils` folder that both this app and MortgagePerformanceTrends use.

el-mapache replied:
I agree, in a follow up PR I will move this to a shared `utils` folder.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/package.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "youth-employment-success",
+  "description": "The consumerfinance.gov/practitioner-resources/resources-youth-employment-programs/transportation-tool/",
+  "homepage": "https://www.consumerfinance.govpractitioner-resources/resources-youth-employment-programs/transportation-tool/",
```
Missing slash in `https://www.consumerfinance.govpractitioner`

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/package.json:
```diff
@@ -0,0 +1,19 @@
+  "private": true,
+  "license": "SEE LICENSE IN TERMS.md",
+  "dependencies": {
+    "keymirror": "0.1.1"
```
Noted earlier… I'd probably refactor to omit keymirror, particularly since it's the only dependency included.

anselmbradford commented on cfgov/youth_employment_success/jinja2/youth-employment-success/route-option-form.html:
```diff
@@ -0,0 +1,92 @@
+<form class="o-yes-route-option">
+  <span class="a-label__heading a-flush__bottom"><b>What type of transportation will you take?</b></span>
+  <span class="a-label__flush a-label_helper__block a-micro-copy">Choose one</span>
+  {% for label in labels|batch(2) %}
```
TRIVIAL: We try to add spaces between the pipe to make it stand out better 

```
{% for label in labels | batch( 2 ) %}
```

Here and below.

el-mapache replied:
Ah good to know. All the examples I saw on the internet seem to have the pipe jammed up against the arguments.

anselmbradford replied:
Yeah for better or worse we tend toward spaces inside things, though it's not totally consistent.

el-mapache commented:
@anselmbradford Thanks for all the comments! You make some good points, I'm going to work through addressing them (and commenting where appropriate).

One thing I've noticed that comes up a few times is using `...` vs directly assigning the property. Ive just gotten into the habit of using explicit reassignments to objects and arrays to avoid mutating them. That matters more in React-land, or places where the object might need to be compared to previous objects to see if they occupy the same memory address. I'm not certain we need that here, so I'll probably go through and mutate the objects where appropriate.

I knew that the `...` operator wasn't supported in IE, but I didn't realize we didn't polyfill it. It looks like we have a pollyfill for object.assign, so I will probably use that.

anselmbradford commented:
> One thing I've noticed that comes up a few times is using ... vs directly assigning the property. Ive just gotten into the habit of using explicit reassignments to objects and arrays to avoid mutating them. That matters more in React-land, or places where the object might need to be compared to previous objects to see if they occupy the same memory address. I'm not certain we need that here, so I'll probably go through and mutate the objects where appropriate.

Just curious, aren't the objects essentially being mutated by re-assigning the original object anyway? Like take:

```
    _state = {
      ..._state,
      [target]: amount
    };
  };
```

`_state` gets replaced with the new object. Is it important in React-land that just prior to the re-assignment of `_state` that it has not been mutated?

el-mapache commented:
@anselmbradford Yep exactly. React might want a reference to the old state for component rendering optimizations. In these components mutation doesn't really matter, although it will when the code is moved to the reducers.

el-mapache commented on cfgov/unprocessed/apps/youth-employment-success/css/main.less:
```diff
@@ -0,0 +1,43 @@
+
+.a-label__flush {
+  &:extend(.a-label);
+  margin-top: 0px;
```
I actually have a lot of questions about layouts, I'm finding it a bit more difficult than I anticipated to match the designs with the current grid / spacing options.

I'll probably hold off on these till we can chat about it.

el-mapache commented on cfgov/unprocessed/apps/youth-employment-success/css/main.less:
```diff
@@ -0,0 +1,43 @@
+  }
+}
+
+.content_line-bold {
```
In this instance, I actually want to preserve the old `comment-line` functionality. The designs call for both a gray dividing line and a black one. Might be better to use a border in the cases of the black line?

anselmbradford replied:
Ah okay. I think mirroring the background color is fine and makes it consistent with `content_line`. One tweak would be that it appears to be a modifier on `content_line`. We use a form of BEM, `some-block_an-element__a-modifier` (see [here](https://github.com/cfpb/development/blob/a3feffa3485f5e2f721f0c6d77d4965ab7d49c94/standards/css.md#naming-conventions). So `content` is the block and `line` is the element. Since you're essentially modifying the color of the same content line, it should be `.content_line__bold`

el-mapache commented on cfgov/unprocessed/apps/youth-employment-success/css/main.less:
```diff
@@ -0,0 +1,43 @@
+    &:extend(.a-label__heading);
+    margin-bottom: 0;
+
+    + input {
```
It worked as expected!

anselmbradford replied:
INNNTeresting

el-mapache commented on cfgov/unprocessed/apps/youth-employment-success/css/main.less:
```diff
@@ -0,0 +1,43 @@
+    margin-bottom: 0;
+
+    + input {
+      margin-top: unit( 15px / 16px, em );
```
I think I may be testing this wrong but this is what I get:

* With a custom font size, and only `px`, I get, of course, a tiny margin in between the text and input.

* With the current code,  and any size font, I see a consistent spacing.

I think bullet two is preferable

anselmbradford replied:
Sounds right! We probably should swap out `16px` for `@base-font-size-px` though to convey that you're setting this to the usual font size and 16px is not a magic number.

el-mapache commented on cfgov/unprocessed/apps/youth-employment-success/js/index.js:
```diff
@@ -0,0 +1,42 @@
+import YesStore from './yes-store';
+import budgetReducer from './reducers/budget-reducer';
+
+window.Expand = Expandable;
```
For debugging purposes, left in by mistake.

el-mapache commented:
@anselmbradford I've addressed most of the comments. The python 2.7 tests are failing, although I'm not sure why. Locally I see an error that the django deps can't be installed.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/budget-form-view.js:
```diff
@@ -0,0 +1,91 @@
+
+  /**
+   *
+   * @param {String} target The piece of state to be updated
```
TRIVIAL: should be `string`

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/budget-form-view.js:
```diff
@@ -0,0 +1,91 @@
+ *
+ * @classdesc Initializes the organism.
+ *
+ * @param {htmlnode} element
```
TRIVIAL: this should be [HTMLElement](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement). We have this wrong elsewhere. The parent interface of HTMLElement is Node.

el-mapache replied:
Will hcange it back

anselmbradford commented:
@el-mapache Looks like this is the backend error:
<img width="667" alt="Screen Shot 2019-08-20 at 2 44 31 PM" src="https://user-images.githubusercontent.com/704760/63375109-21ea0000-c359-11e9-9f0a-40a77c140aec.png">

el-mapache commented:
@anselmbradford I was definitely not looking for anything like that 😄

anselmbradford commented:
> Re: my use of the ... operator, it seems that running eslint with the --fix flag automatically replaces Object.assign with the ... operator. Does this mean it is safe to use? Or does it mean that I need to expressly do a for loop with hasOwnProperty checks?

Hmm, right you are. [Setting here](https://github.com/cfpb/cfgov-refresh/blob/master/.eslintrc#L1009). I guess we just hadn't run into the need for Object.assign elsewhere in the codebase yet where it mattered (I see it used in some of our build and an analytics log script, which both don't care about IE). Actually, there is a use in owning-a-home, which I need to check on in IE11. We have a few options here:

1. Adjust the ESLint config to be a warning.
2. Add an `.eslintrc` config file to your app folder that overrides the global one for this setting.
3. Use `Object.assign` without an object literal as the first argument (see "correct code for this rule" in https://eslint.org/docs/rules/prefer-object-spread) and include the polyfill.
4. Follow something like https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/unprocessed/apps/owning-a-home/js/explore-rates/data-loader.js#L37-L43 (including the code comment so we can refactor it when IE11 support is dropped.

Any preferred option of those for you?

el-mapache commented:
@anselmbradford I'l go with number 4 I think.

el-mapache commented:
@anselmbradford Updated!

el-mapache commented:
There param in there because it is used in the text input. We need a way to
map the state to the name of the field. Since most of the fields have
autogenerated names with a random number, I wanted a simple way to get that
mapping.

The function now supplies an object as the argument, so we can just pull
the name out for the fields that need it.

On Wed, Aug 21, 2019 at 8:23 AM Ans <notifications@github.com> wrote:

> *@anselmbradford* commented on this pull request.
> ------------------------------
>
> In cfgov/unprocessed/apps/youth-employment-success/js/route-option-view.js
> <https://github.com/cfpb/cfgov-refresh/pull/5176#discussion_r316248217>:
>
> > +
> +  /**
> +   * Updates form state from child <input text> nodes
> +   * @param {String} name Identifier of the piece of state to update
> +   * @param {Object} event DOM event object
> +   */
> +  function _setQuestionResponse( name, event ) {
> +    state = {
> +      ...state,
> +      [name]: event.target.value
> +    };
> +  }
> +
> +  /**
> +   * Updates state from child <input checkbox> nodes
> +   * @param {*} _ unused
>
> I mean why's the param there at all?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cfpb/cfgov-refresh/pull/5176?email_source=notifications&email_token=AAK3EGGSFYMI7Z5KDSZ262LQFVMVNA5CNFSM4IMK4PT2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCCIA5XI#discussion_r316248217>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAK3EGHWDYBA36A5IIDD2ELQFVMVNANCNFSM4IMK4PTQ>
> .
>

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/budget-form-view.js:
```diff
@@ -0,0 +1,91 @@
+  /**
+   *
+   * @param {string} target The piece of state to be updated
+   * @returns {function} A function that accepts an event and updates
```
Trivial: technically this should be `Function` since it's not a primitive.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/index.js:
```diff
@@ -0,0 +1,41 @@
+expandables.reverse().forEach( expandable => {
+  expandable.element.querySelector( '.o-expandable_target' ).click();
+
+  /* why doesn't this work properly?
```
Possibly prefix this comment with `TODO: ` so it shows up in code search.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/yes-store.js:
```diff
@@ -0,0 +1,17 @@
+    super( middlewares );
+
+    this.reducer = reducer;
+    this.state = this.reduce( UNDEFINED );
```
If you're using undefined only once, probably just have `let UNDEFINED;` above this line and get rid of the import at the top.

anselmbradford replied:
Here and elsewhere. We used to have `UNDEFINED` in a utils but ended up just defining it where it was needed as it was only used once usually.

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/input-view.js:
```diff
@@ -0,0 +1,110 @@
+ * Get the correct DOM node this view controls
+ * @param {node} element dom node representing either the input this
+ *  view controls or it's parent
+ * @param {*} type the kind of input (text, checkbox, etc)
```
Would this be of type string?

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/util.js:
```diff
@@ -0,0 +1,110 @@
+ * Throw an error if the reducer processing the current action doesn't return
+ * a value
+ * @param {object} nextState The state object just returned by a reducer
+ */
```
Comment could have `* @throws {Error} If the reducer processing the current action doesn't return a value.`

el-mapache replied:
Oh cool, my bad for not verifying if there was a `@throw` declaration

anselmbradford commented on cfgov/unprocessed/apps/youth-employment-success/js/util.js:
```diff
@@ -0,0 +1,110 @@
+  return hasChanged ? finalState : state;
+};
+
+export const combineReducers = reducers => {
```
Would it make sense to define `const combineReducers = …` here and then at the end of the file have `export combineReducers;`, so that all exports are together?

el-mapache replied:
yes, I will normalize that for sure in my next PR!
